### ISSUE TITLE
resolveIdentifier regression issue after upgrading from 3.0.2 to 3.2.2

### ISSUE THREAD

**Issue by RainerAtSpirit** [mobxjs/mobx-state-tree#1019]
I have:

* [ ] A conceptual question. 
  * [ ] I've checked documentation and searched for existing issues
  * [ ] I tried the [spectrum channel](https://spectrum.chat/?t=dad48299-3dfc-4e10-b6da-9af1e39498a3) first
* [x] I think something is not working as it should.
  * [x] I've checked documentation and searched for existing issues
  * [ ] I've made sure your project is based on the latest MST version
  * [x] Fork [this](https://codesandbox.io/s/jlnmjqr7xv) code sandbox or another minimal reproduction. 
  * [x] Describe expected behavior
MST3.0.2 `resolveIdentifier` working:
https://codesandbox.io/s/24zrq4n9n0 (models.CommentStore.js, line 15...)
```javascript
 changeSelectedItem(uid: string) {
      // Working in MST3.0.2
      const item = resolveIdentifier(CommentModel, self.items, uid);
      console.log("item", item && item.toJSON());
      const item2 = self.items.find(i => i.uid === uid);
      console.log("item2", item2 && item2.toJSON());

      self.selectedItem = item;
    }
```
  * [x] Describe observed behavior

Regression issue: After upgrading to MST3.2.2 `resolveIdentifier` returns undefined

https://codesandbox.io/s/294ylryp0

* [ ] Feature request
   * [ ] Describe the feature and why you feel your feature needs to be generically solved inside MST
   * [ ] Are you willing to (attempt) a PR?
   
_Not following the above template might result in your issue being closed without further notice_

**Comment by xaviergonz** [comment]
I took a look into it, it seems your id generator does not pass the test for the TOptionalId refinement

**Comment by xaviergonz** [comment]
nevermind, it seems to be about the identifier being optional with a function, looking further into it

**Comment by xaviergonz** [comment]
unit test
```ts
test("#1019", () => {
    function randomUuid() {
        return "1bbc17fd-2068-4cc9-aff0-82d37c9cdf20"
    }

    const CommentModel = types.model("CommentModel", {
        uid: types.optional(types.identifier, randomUuid)
    })

    const CommentStore = types
        .model("CommentStore", {
            items: types.array(CommentModel)
        })
        .actions(self => {
            function test() {
                const comment = CommentModel.create({})
                expect(comment.uid).toBe(randomUuid())

                self.items.push(comment)
                const item = resolveIdentifier(CommentModel, self.items, comment.uid)
                const item2 = self.items.find(i => i.uid === comment.uid)
                expect(item).toBe(item2)
            }

            return {
                test
            }
        })

    CommentStore.create({}).test()
})
````

**Comment by galdebert** [comment]
I had a similar (I think) issue today, a minimal repro is
```
  const Item = types
    .model({
      id: types.optional(types.identifier, 'dummykey'),
    })
  console.log(getIdentifier(Item.create())) // undefined, should be 'dummykey'
```

**Comment by xaviergonz** [comment]
The PR I just opened should fix both issues

**Comment by xaviergonz** [comment]
should be fixed in the recently released 3.4.0
